astropy/modeling/separable.py

```diff
@@ -0,0 +1,293 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+"""
+Functions to determine if a model is separable, i.e.
+if the model outputs are independent.
+
+It analyzes ``n_inputs``, ``n_outputs`` and the operators
+in a compound model by stepping through the transforms
+and creating a ``coord_matrix`` of shape (``n_outputs``, ``n_inputs``).
+
+
+Each modeling operator is represented by a function which
+takes two simple models (or two ``coord_matrix`` arrays) and
+returns an array of shape (``n_outputs``, ``n_inputs``).
+
+"""
+
+import numpy as np
+
+from .core import Model, _CompoundModel, ModelDefinitionError
+from .mappings import Mapping
+from .models import Shift, Scale, Rotation2D, Polynomial2D
+
+
+__all__ = ["is_separable"]
+
+
+def is_separable(transform):
+    """
+    A separability test for the outputs of a transform.
+
+    Parameters
+    ----------
+    transform : `~astropy.modeling.core.Model`
+        A (compound) model.
+
+    Returns
+    -------
+    is_separable : ndarray
+        A boolean array with size ``transform.n_outputs`` where
+        each element indicates whether the output is independent
+        and the result of a separable transform.
+
+     Examples
+    --------
+    >>> is_separable(Shift(1) & Shift(2) | Scale(1) & Scale(2))
```

Comment:
I have a feeling that this won't pass `doctest` due to missing imports.

Reply:
I guess this is more of a question for @bsipocz -- How did this pass the doctest? I thought we test codes inside docstring, right? If so, is a docstring from hidden function (below) tested?

Reply:
It seems that doctest has access all the imports of a given file, so has access to both ``is_separable`` as well as the models used here. I would still suggest to have full working examples in the example section, with doing all the necessary imports.